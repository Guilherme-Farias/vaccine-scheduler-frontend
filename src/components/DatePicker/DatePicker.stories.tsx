import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import 'react-datepicker/dist/react-datepicker.css';
import DatePicker from '.';

export default {
  title: 'Form/DatePicker',
  component: DatePicker,
  argTypes: {
    label: {
      description: 'Label text',
      type: 'string',
    },
    error: {
      type: 'string',
      description: 'Informs if the field has error',
    },
    disabled: {
      type: 'boolean',
      description: 'Informs if the field is disabled',
    },
    readOnly: {
      type: 'boolean',
      description: 'Informs if the field is readOnly',
    },
    style: {
      description: 'React CSS Properties',
      control: { type: '' },
    },
  },
} as ComponentMeta<typeof DatePicker>;

const Template: ComponentStory<typeof DatePicker> = args => (
  <div
    style={{
      display: 'grid',
      justifyContent: 'center',
      alignContent: 'center',
    }}
  >
    <h1>For better documentation go to:</h1>
    <ul>
      <li>
        <a
          href="https://www.npmjs.com/package/react-datepicker"
          target="_blank"
          rel="noreferrer"
        >
          React DatePicker npmjs
        </a>
      </li>
      <li>
        <a
          href="https://github.com/Hacker0x01/react-datepicker"
          target="_blank"
          rel="noreferrer"
        >
          React DatePicker Github
        </a>
      </li>
      <li>
        <a href="https://reactdatepicker.com/" target="_blank" rel="noreferrer">
          React DatePicker Page
        </a>
      </li>
    </ul>
    <DatePicker {...args} />
  </div>
);

export const Default = Template.bind({});
Default.args = {
  label: 'Select the date',
  error: '',
  placeholderText: 'dd/mm/aaaa',
  onChange: () => console.log('onChange'),
};
