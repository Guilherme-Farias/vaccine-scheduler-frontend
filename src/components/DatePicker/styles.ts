import styled, { css, DefaultTheme } from 'styled-components/macro';
import ReactDatePicker from 'react-datepicker';
import { FormGroup } from 'components/FormControls';

export const DatePicker = styled(ReactDatePicker)`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-size: ${theme.font.sizes.md};
    outline: none;

    width: 100%;
    height: ${theme.spacings['5xl']};
    background: ${theme.colors.white};
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xs} ${theme.spacings.sm};
    border: 1px solid;
    border-color: ${theme.colors['gray-500']};

    transition: border-color 0.2s ease-in-out, box-shadow 0.2s ease-in-out;
    &:hover {
      border-color: ${theme.colors['gray-600']};
      box-shadow: ${theme.colors['gray-600']} 0px 0px 0px 1px;
    }
    &:focus {
      border-color: ${theme.colors.blue};
      box-shadow: ${theme.colors.blue} 0px 0px 0px 1px;
    }
  `}
`;

const styledFormGroupModifiers = {
  isDisabled: (theme: DefaultTheme) => css`
    &,
    * {
      cursor: not-allowed;
    }

    ${DatePicker} {
      color: ${theme.colors['gray-800']};
      background-color: ${theme.colors['gray-300']};
      &::placeholder {
        color: currentColor;
      }
    }
  `,
  isInvalid: (theme: DefaultTheme) => css`
    ${DatePicker} {
      border-color: ${theme.colors.red};
      &:focus {
        box-shadow: ${theme.colors.red} 0px 0px 0px 1px;
      }
    }
  `,
  isReadOnly: (theme: DefaultTheme) => css`
    ${DatePicker} {
      border-color: ${theme.colors['gray-600']};
      background-color: ${theme.colors['gray-300']};
      &:focus {
        border-color: ${theme.colors.blue};
        box-shadow: ${theme.colors.blue} 0px 0px 0px 1px;
      }
    }
  `,
};

export const StyledFormGroup = styled(FormGroup)`
  ${({ theme, isDisabled, isInvalid, isReadOnly }) => css`
    ${!!isDisabled && styledFormGroupModifiers.isDisabled(theme)};
    ${!!isInvalid && styledFormGroupModifiers.isInvalid(theme)};
    ${!!isReadOnly && styledFormGroupModifiers.isReadOnly(theme)};
  `}
`;
