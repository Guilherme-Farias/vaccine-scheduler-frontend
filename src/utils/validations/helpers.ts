import Joi from 'joi';
import { IParamErrors } from 'types/global';

export function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: IParamErrors = {};

  if (objError.error) {
    objError.error.details.forEach(err => {
      errors[err.path.join('.')] = err.message;
    });
  }

  return errors;
}
