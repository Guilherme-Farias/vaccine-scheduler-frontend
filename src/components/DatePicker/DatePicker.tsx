import React from 'react';

import { ReactDatePickerProps } from 'react-datepicker';

import { FormLabel, FormErrorMessage } from 'components/FormControls';

import * as S from './styles';

export type DatePickerProps = {
  label?: string;
  error?: string;
  style?: React.CSSProperties;
} & ReactDatePickerProps;

const DatePicker: React.FC<DatePickerProps> = ({
  name,
  label = '',
  error = '',
  disabled,
  readOnly,
  className,
  style,
  ...props
}: DatePickerProps) => (
  <S.StyledFormGroup
    isDisabled={disabled}
    isInvalid={!!error}
    isReadOnly={readOnly}
    className={className}
    style={style}
  >
    {!!label && (
      <FormLabel htmlFor={name} isDisabled={disabled} isInvalid={!!error}>
        {label}
      </FormLabel>
    )}
    <S.DatePicker
      name={name}
      disabled={disabled}
      readOnly={readOnly}
      {...(label ? { id: name } : {})}
      {...props}
    />
    {!!error && <FormErrorMessage>{error}</FormErrorMessage>}
  </S.StyledFormGroup>
);

export default DatePicker;
