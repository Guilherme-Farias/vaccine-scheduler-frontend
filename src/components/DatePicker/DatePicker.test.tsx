import { render, screen, userEvent } from 'utils/tests/helpers';

import DatePicker from '.';

const onChangeSpy = jest.fn();
describe('<DatePicker />', () => {
  it('should render with Label', () => {
    render(<DatePicker label="Test" name="test" onChange={onChangeSpy} />);

    expect(screen.getByLabelText(/test/i)).toBeInTheDocument();
  });

  it('should render with Label', () => {
    render(<DatePicker onChange={onChangeSpy} />);

    expect(screen.queryByLabelText(/test/i)).not.toBeInTheDocument();
  });

  it('should render with error', () => {
    const { container } = render(
      <DatePicker
        onChange={onChangeSpy}
        label="Test"
        name="test"
        error="Error message"
      />,
    );

    expect(screen.getByText('Error message')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not changes its value when disabled', async () => {
    render(
      <DatePicker
        onChange={onChangeSpy}
        label="Test"
        name="test"
        error="Error message"
        disabled
      />,
    );

    const input = screen.getByRole('textbox');
    expect(input).toBeDisabled();

    const text = 'This is my new text';
    await userEvent.type(input, text);

    expect(input).not.toHaveValue(text);
  });

  it('should not changes its value when read only', async () => {
    render(
      <DatePicker
        onChange={onChangeSpy}
        label="Test"
        name="test"
        error="Error message"
        readOnly
      />,
    );

    const input = screen.getByRole('textbox');
    expect(input).toHaveAttribute('readOnly');

    const text = 'This is my new text';
    await userEvent.type(input, text);

    expect(input).not.toHaveValue(text);
  });

  it('should be accessible by tab', () => {
    render(<DatePicker onChange={onChangeSpy} label="Test" name="test" />);

    const input = screen.getByLabelText(/test/i);
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  });

  it('should not accessible by tab when disabled', () => {
    render(
      <DatePicker onChange={onChangeSpy} label="Test" name="test" disabled />,
    );

    const input = screen.getByLabelText(/test/i);
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).not.toHaveFocus();
  });
});
