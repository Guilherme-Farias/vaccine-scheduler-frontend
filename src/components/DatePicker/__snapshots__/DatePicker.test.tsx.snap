// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> should render with error 1`] = `
.c1 {
  cursor: pointer;
  display: inline-block;
  color: #171923;
  font-size: 1rem;
  font-weight: 500;
  margin-bottom: 0.5rem;
  color: #DC2626;
}

.c4 {
  color: #DC2626;
  font-size: 0.875rem;
}

.c3 {
  color: #171923;
  font-size: 1rem;
  outline: none;
  width: 100%;
  height: 2.5rem;
  background: #FFFFFF;
  border-radius: 0.625rem;
  padding: 0.5rem 0.75rem;
  border: 1px solid;
  border-color: #6b7280;
  -webkit-transition: border-color 0.2s ease-in-out,box-shadow 0.2s ease-in-out;
  transition: border-color 0.2s ease-in-out,box-shadow 0.2s ease-in-out;
}

.c3:hover {
  border-color: #4b5563;
  box-shadow: #4b5563 0px 0px 0px 1px;
}

.c3:focus {
  border-color: #2563eb;
  box-shadow: #2563eb 0px 0px 0px 1px;
}

.c0 .c2 {
  border-color: #DC2626;
}

.c0 .c2:focus {
  box-shadow: #DC2626 0px 0px 0px 1px;
}

<div
  class=" c0"
  role="group"
>
  <label
    class="c1"
    for="test"
  >
    Test
  </label>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        class="c2 c3"
        id="test"
        name="test"
        type="text"
        value=""
      />
    </div>
  </div>
  <span
    class="c4"
    role="alert"
  >
    Error message
  </span>
</div>
`;
